File: frontend/src/infrastructure/resilience/circuitBreaker.ts
/**
 * Circuit Breaker Pattern Implementation
 *
 * Infrastructure layer component for resilience patterns.
 * Prevents cascading failures by stopping requests to failing services
 * and giving them time to recover.
 */

import { dispatchEvent, AppEventType } from '../events'

export interface CircuitBreakerConfig {
  failureThreshold: number  // Number of failures before opening circuit
  resetTimeout: number      // Time in ms before attempting to close circuit
  monitoringWindow?: number // Time window for tracking failures (default: same as resetTimeout)
}

export type CircuitState = 'closed' | 'open' | 'half-open'

export interface CircuitBreakerStats {
  state: CircuitState
  failureCount: number
  successCount: number
  lastFailureTime: number | null
  lastStateChange: number
}

/**
 * CircuitBreaker implementation with state management
 *
 * States:
 * - CLOSED: Normal operation, requests pass through
 * - OPEN: Circuit is open, requests fail immediately
 * - HALF_OPEN: Testing if service has recovered, limited requests allowed
 */
export class CircuitBreaker {
  private failureCount = 0
  private successCount = 0
  private lastFailureTime: number | null = null
  private lastStateChange: number = Date.now()
  private state: CircuitState = 'closed'
  private readonly config: Required<CircuitBreakerConfig>

  constructor(config: CircuitBreakerConfig) {
    this.config = {
      ...config,
      monitoringWindow: config.monitoringWindow ?? config.resetTimeout,
    }
  }

  /**
   * Execute a function with circuit breaker protection
   */
  async execute<T>(fn: () => Promise<T>): Promise<T> {
    // Check if enough time has passed to transition from OPEN to HALF_OPEN
    if (this.state === 'open') {
      const timeSinceLastFailure = Date.now() - (this.lastFailureTime ?? 0)

      if (timeSinceLastFailure > this.config.resetTimeout) {
        this.transitionToHalfOpen()
      } else {
        throw new CircuitBreakerError(
          'Circuit breaker is open - service temporarily unavailable',
          this.getStats()
        )
      }
    }

    try {
      const result = await fn()
      this.onSuccess()
      return result
    } catch (error) {
      this.onFailure()
      throw error
    }
  }

  /**
   * Handle successful execution
   */
  private onSuccess(): void {
    this.successCount++

    if (this.state === 'half-open') {
      // Success in half-open state means service recovered
      this.transitionToClosed()
    } else if (this.state === 'closed') {
      // Reset failure count on success in closed state
      this.failureCount = 0
    }
  }

  /**
   * Handle failed execution
   */
  private onFailure(): void {
    this.failureCount++
    this.lastFailureTime = Date.now()

    if (this.state === 'half-open') {
      // Failure in half-open state means service still unhealthy
      this.transitionToOpen()
    } else if (
      this.state === 'closed' &&
      this.failureCount >= this.config.failureThreshold
    ) {
      // Too many failures in closed state
      this.transitionToOpen()
    }
  }

  /**
   * Transition to CLOSED state (normal operation)
   */
  private transitionToClosed(): void {
    this.state = 'closed'
    this.failureCount = 0
    this.successCount = 0
    this.lastStateChange = Date.now()
    console.info('[CircuitBreaker] Transitioned to CLOSED state - service healthy')

    // Dispatch event
    dispatchEvent(AppEventType.API_CIRCUIT_BREAKER_CLOSED, {
      recoveryTime: Date.now() - this.lastStateChange,
    }).catch(console.error)
  }

  /**
   * Transition to OPEN state (blocking requests)
   */
  private transitionToOpen(): void {
    this.state = 'open'
    this.lastStateChange = Date.now()
    console.warn(
      `[CircuitBreaker] Transitioned to OPEN state - ${this.failureCount} failures detected`
    )

    // Dispatch event
    dispatchEvent(AppEventType.API_CIRCUIT_BREAKER_OPEN, {
      failureCount: this.failureCount,
    }).catch(console.error)
  }

  /**
   * Transition to HALF_OPEN state (testing recovery)
   */
  private transitionToHalfOpen(): void {
    this.state = 'half-open'
    this.failureCount = 0
    this.successCount = 0
    this.lastStateChange = Date.now()
    console.info('[CircuitBreaker] Transitioned to HALF_OPEN state - testing service recovery')
  }

  /**
   * Get current circuit breaker statistics
   */
  getStats(): CircuitBreakerStats {
    return {
      state: this.state,
      failureCount: this.failureCount,
      successCount: this.successCount,
      lastFailureTime: this.lastFailureTime,
      lastStateChange: this.lastStateChange,
    }
  }

  /**
   * Get current state
   */
  getState(): CircuitState {
    return this.state
  }

  /**
   * Manually reset circuit breaker (for testing/admin purposes)
   */
  reset(): void {
    this.transitionToClosed()
  }
}

/**
 * Custom error for circuit breaker open state
 */
export class CircuitBreakerError extends Error {
  constructor(
    message: string,
    public stats: CircuitBreakerStats
  ) {
    super(message)
    this.name = 'CircuitBreakerError'
  }
}

File: src/codetoreum/infrastructure/resilience/circuit_breaker.py
"""Production circuit breaker implementation.

Provides circuit breaker pattern with CLOSED/OPEN/HALF_OPEN states.
"""

import asyncio
import time
from collections.abc import Callable
from datetime import datetime
from typing import TypeVar

from .exceptions import CircuitBreakerOpenError
from .interfaces import CircuitBreakerStats, CircuitState, ICircuitBreaker

T = TypeVar("T")


class CircuitBreaker(ICircuitBreaker):
    """
    Circuit breaker with half-open state for recovery testing.

    State transitions:
    - CLOSED -> OPEN: After failure_threshold consecutive failures
    - OPEN -> HALF_OPEN: After timeout_seconds elapsed
    - HALF_OPEN -> CLOSED: After success_threshold consecutive successes
    - HALF_OPEN -> OPEN: On any failure
    """

    def __init__(
        self,
        failure_threshold: int = 5,
        timeout_seconds: float = 60,
        success_threshold: int = 2,
        expected_exceptions: tuple[type[Exception], ...] = (Exception,),
    ):
        """
        Initialize circuit breaker.

        Args:
            failure_threshold: Failures before opening circuit
            timeout_seconds: Time before attempting recovery (HALF_OPEN)
            success_threshold: Successes needed to close circuit from HALF_OPEN
            expected_exceptions: Exceptions that trigger circuit breaker
        """
        self.failure_threshold = failure_threshold
        self.timeout_seconds = timeout_seconds
        self.success_threshold = success_threshold
        self.expected_exceptions = expected_exceptions

        self._state = CircuitState.CLOSED
        self._failure_count = 0
        self._success_count = 0
        self._last_failure_time: float | None = None
        self._last_success_time: float | None = None

        # Statistics
        self._total_calls = 0
        self._total_failures = 0
        self._total_successes = 0

        self._lock = asyncio.Lock()

    async def call(self, operation: Callable[..., T], operation_name: str, *args, **kwargs) -> T:
        """Execute operation with circuit breaker protection."""
        async with self._lock:
            self._total_calls += 1

            # Check if circuit should transition to HALF_OPEN
            if self._state == CircuitState.OPEN:
                if self._should_attempt_reset():
                    self._state = CircuitState.HALF_OPEN
                    self._success_count = 0
                else:
                    message = (
                        f"Circuit breaker open for {operation_name}. Will retry in {self._time_until_retry():.1f}s"
                    )
                    raise CircuitBreakerOpenError(message, retry_after_seconds=self._time_until_retry())

        # Execute operation (outside lock to avoid holding during I/O)
        try:
            result = await operation(*args, **kwargs)

            # Handle success
            async with self._lock:
                self._on_success()

            return result

        except self.expected_exceptions:
            # Handle failure
            async with self._lock:
                self._on_failure()
            raise

    def _should_attempt_reset(self) -> bool:
        """Check if enough time has passed to attempt recovery."""
        if self._last_failure_time is None:
            return True
        return time.time() - self._last_failure_time >= self.timeout_seconds

    def _time_until_retry(self) -> float:
        """Calculate time until next retry attempt."""
        if self._last_failure_time is None:
            return 0
        elapsed = time.time() - self._last_failure_time
        return max(0, self.timeout_seconds - elapsed)

    def _on_success(self) -> None:
        """Handle successful operation."""
        self._last_success_time = time.time()
        self._total_successes += 1

        if self._state == CircuitState.HALF_OPEN:
            self._success_count += 1
            if self._success_count >= self.success_threshold:
                # Recovered! Close the circuit
                self._state = CircuitState.CLOSED
                self._failure_count = 0
        elif self._state == CircuitState.CLOSED:
            # Reset failure count on success
            self._failure_count = 0

    def _on_failure(self) -> None:
        """Handle failed operation."""
        self._last_failure_time = time.time()
        self._total_failures += 1
        self._failure_count += 1

        if self._state == CircuitState.HALF_OPEN:
            # Failed during recovery - reopen circuit
            self._state = CircuitState.OPEN
        elif self._state == CircuitState.CLOSED:
            # Check if we should open circuit
            if self._failure_count >= self.failure_threshold:
                self._state = CircuitState.OPEN

    def get_state(self) -> CircuitState:
        """Get current circuit state."""
        return self._state

    def get_stats(self) -> CircuitBreakerStats:
        """Get statistics."""
        return CircuitBreakerStats(
            state=self._state,
            failure_count=self._failure_count,
            success_count=self._success_count,
            last_failure_time=datetime.fromtimestamp(self._last_failure_time) if self._last_failure_time else None,
            last_success_time=datetime.fromtimestamp(self._last_success_time) if self._last_success_time else None,
            total_calls=self._total_calls,
            total_failures=self._total_failures,
            total_successes=self._total_successes,
        )

    def reset(self) -> None:
        """Reset to initial state."""
        self._state = CircuitState.CLOSED
        self._failure_count = 0
        self._success_count = 0
        self._last_failure_time = None
        self._last_success_time = None

    def force_open(self) -> None:
        """Manually open circuit."""
        self._state = CircuitState.OPEN
        self._last_failure_time = time.time()

    def force_close(self) -> None:
        """Manually close circuit."""
        self._state = CircuitState.CLOSED
        self._failure_count = 0
